File: src-tauri/src/commands/library_storage.rs
use serde_json::Value;
use std::path::PathBuf;
use tokio::fs;

const LIBRARY_DIR: &str = "library";
const INDEX_FILE: &str = "index.json";
const ARTIFACTS_DIR: &str = "artifacts";

// Helper to get the library directory
fn get_library_dir(app_handle: &tauri::AppHandle) -> Result<PathBuf, String> {
    let app_data_dir = app_handle
        .path()
        .app_data_dir()
        .map_err(|e| format!("Failed to get app data directory: {}", e))?;
    Ok(app_data_dir.join(LIBRARY_DIR))
}

// Helper to get the index file path
fn get_index_path(app_handle: &tauri::AppHandle) -> Result<PathBuf, String> {
    Ok(get_library_dir(app_handle)?.join(INDEX_FILE))
}

// Helper to get the artifacts directory
fn get_artifacts_dir(app_handle: &tauri::AppHandle) -> Result<PathBuf, String> {
    Ok(get_library_dir(app_handle)?.join(ARTIFACTS_DIR))
}

// ============================================
// Library Index Commands
// ============================================

/// Get the library index (list of all artifacts)
#[tauri::command]
pub async fn get_library_index(app_handle: tauri::AppHandle) -> Result<String, String> {
    let index_path = get_index_path(&app_handle)?;

    // If file doesn't exist, return default structure
    if !index_path.exists() {
        let default = serde_json::json!({
            "version": 1,
            "lastUpdated": chrono::Utc::now().to_rfc3339(),
            "artifacts": []
        });
        return Ok(default.to_string());
    }

    fs::read_to_string(&index_path)
        .await
        .map_err(|e| format!("Failed to read library index: {}", e))
}

/// Save the library index
#[tauri::command]
pub async fn save_library_index(
    app_handle: tauri::AppHandle,
    index: String,
) -> Result<(), String> {
    let library_dir = get_library_dir(&app_handle)?;
    let index_path = get_index_path(&app_handle)?;

    // Create directory if it doesn't exist
    fs::create_dir_all(&library_dir)
        .await
        .map_err(|e| format!("Failed to create library directory: {}", e))?;

    // Validate JSON
    let _: Value =
        serde_json::from_str(&index).map_err(|e| format!("Invalid index JSON: {}", e))?;

    // Write index
    fs::write(&index_path, &index)
        .await
        .map_err(|e| format!("Failed to write library index: {}", e))?;

    Ok(())
}

// ============================================
// Artifact Commands
// ============================================

/// Get a specific artifact by ID
#[tauri::command]
pub async fn get_artifact(
    app_handle: tauri::AppHandle,
    artifact_id: String,
) -> Result<String, String> {
    let artifacts_dir = get_artifacts_dir(&app_handle)?;
    let artifact_path = artifacts_dir.join(format!("{}.json", artifact_id));

    if !artifact_path.exists() {
        return Err(format!("Artifact not found: {}", artifact_id));
    }

    fs::read_to_string(&artifact_path)
        .await
        .map_err(|e| format!("Failed to read artifact: {}", e))
}

/// Save an artifact (create or update)
#[tauri::command]
pub async fn save_artifact(
    app_handle: tauri::AppHandle,
    artifact: String,
) -> Result<(), String> {
    let library_dir = get_library_dir(&app_handle)?;
    let artifacts_dir = get_artifacts_dir(&app_handle)?;
    let index_path = get_index_path(&app_handle)?;

    // Create directories if they don't exist
    fs::create_dir_all(&artifacts_dir)
        .await
        .map_err(|e| format!("Failed to create artifacts directory: {}", e))?;

    // Parse the incoming artifact
    let artifact_value: Value =
        serde_json::from_str(&artifact).map_err(|e| format!("Invalid artifact JSON: {}", e))?;

    let artifact_id = artifact_value
        .get("artifactId")
        .and_then(|v| v.as_str())
        .ok_or("Artifact must have an artifactId")?;

    // Save the full artifact to its own file
    let artifact_path = artifacts_dir.join(format!("{}.json", artifact_id));
    fs::write(&artifact_path, &artifact)
        .await
        .map_err(|e| format!("Failed to write artifact: {}", e))?;

    // Update the index
    let mut index: Value = if index_path.exists() {
        let content = fs::read_to_string(&index_path)
            .await
            .map_err(|e| format!("Failed to read library index: {}", e))?;
        serde_json::from_str(&content).unwrap_or_else(|_| {
            serde_json::json!({
                "version": 1,
                "lastUpdated": chrono::Utc::now().to_rfc3339(),
                "artifacts": []
            })
        })
    } else {
        serde_json::json!({
            "version": 1,
            "lastUpdated": chrono::Utc::now().to_rfc3339(),
            "artifacts": []
        })
    };

    // Create index entry (metadata only, no HTML content)
    let index_entry = serde_json::json!({
        "artifactId": artifact_value.get("artifactId"),
        "projectId": artifact_value.get("projectId"),
        "jobId": artifact_value.get("jobId"),
        "type": artifact_value.get("type"),
        "title": artifact_value.get("title"),
        "grade": artifact_value.get("grade"),
        "subject": artifact_value.get("subject"),
        "objectiveTags": artifact_value.get("objectiveTags"),
        "designPackId": artifact_value.get("designPackId"),
        "createdAt": artifact_value.get("createdAt"),
    });

    // Update artifacts array in index
    if let Some(artifacts) = index.get_mut("artifacts") {
        if let Some(arr) = artifacts.as_array_mut() {
            // Remove existing entry with same ID
            arr.retain(|a| a.get("artifactId").and_then(|v| v.as_str()) != Some(artifact_id));
            // Add new entry
            arr.push(index_entry);
        }
    }

    // Update lastUpdated
    if let Some(obj) = index.as_object_mut() {
        obj.insert(
            "lastUpdated".to_string(),
            Value::String(chrono::Utc::now().to_rfc3339()),
        );
    }

    // Write index
    let index_content = serde_json::to_string_pretty(&index)
        .map_err(|e| format!("Failed to serialize index: {}", e))?;
    fs::write(&index_path, index_content)
        .await
        .map_err(|e| format!("Failed to write library index: {}", e))?;

    Ok(())
}

/// Delete an artifact
#[tauri::command]
pub async fn delete_artifact(
    app_handle: tauri::AppHandle,
    artifact_id: String,
) -> Result<(), String> {
    let artifacts_dir = get_artifacts_dir(&app_handle)?;
    let index_path = get_index_path(&app_handle)?;
    let artifact_path = artifacts_dir.join(format!("{}.json", artifact_id));

    // Delete artifact file if it exists
    if artifact_path.exists() {
        fs::remove_file(&artifact_path)
            .await
            .map_err(|e| format!("Failed to delete artifact file: {}", e))?;
    }

    // Update index
    if index_path.exists() {
        let content = fs::read_to_string(&index_path)
            .await
            .map_err(|e| format!("Failed to read library index: {}", e))?;
        let mut index: Value = serde_json::from_str(&content).unwrap_or_else(|_| {
            serde_json::json!({
                "version": 1,
                "lastUpdated": chrono::Utc::now().to_rfc3339(),
                "artifacts": []
            })
        });

        if let Some(artifacts) = index.get_mut("artifacts") {
            if let Some(arr) = artifacts.as_array_mut() {
                arr.retain(|a| a.get("artifactId").and_then(|v| v.as_str()) != Some(&artifact_id));
            }
        }

        if let Some(obj) = index.as_object_mut() {
            obj.insert(
                "lastUpdated".to_string(),
                Value::String(chrono::Utc::now().to_rfc3339()),
            );
        }

        let index_content = serde_json::to_string_pretty(&index)
            .map_err(|e| format!("Failed to serialize index: {}", e))?;
        fs::write(&index_path, index_content)
            .await
            .map_err(|e| format!("Failed to write library index: {}", e))?;
    }

    Ok(())
}

/// Search artifacts with filters
#[tauri::command]
pub async fn search_artifacts(
    app_handle: tauri::AppHandle,
    query: String,
) -> Result<String, String> {
    let index_path = get_index_path(&app_handle)?;

    if !index_path.exists() {
        return Ok("[]".to_string());
    }

    // Parse query
    let query_value: Value =
        serde_json::from_str(&query).map_err(|e| format!("Invalid query JSON: {}", e))?;

    // Read index
    let content = fs::read_to_string(&index_path)
        .await
        .map_err(|e| format!("Failed to read library index: {}", e))?;
    let index: Value = serde_json::from_str(&content).unwrap_or_else(|_| {
        serde_json::json!({
            "version": 1,
            "lastUpdated": chrono::Utc::now().to_rfc3339(),
            "artifacts": []
        })
    });

    let artifacts = index.get("artifacts").and_then(|v| v.as_array());
    if artifacts.is_none() {
        return Ok("[]".to_string());
    }
    let artifacts = artifacts.unwrap();

    // Apply filters
    let filtered: Vec<&Value> = artifacts
        .iter()
        .filter(|artifact| {
            // Project ID filter
            if let Some(project_id) = query_value.get("projectId").and_then(|v| v.as_str()) {
                if artifact.get("projectId").and_then(|v| v.as_str()) != Some(project_id) {
                    return false;
                }
            }

            // Grade filter
            if let Some(grade) = query_value.get("grade").and_then(|v| v.as_str()) {
                if artifact.get("grade").and_then(|v| v.as_str()) != Some(grade) {
                    return false;
                }
            }

            // Subject filter
            if let Some(subject) = query_value.get("subject").and_then(|v| v.as_str()) {
                if artifact.get("subject").and_then(|v| v.as_str()) != Some(subject) {
                    return false;
                }
            }

            // Type filter
            if let Some(artifact_type) = query_value.get("type").and_then(|v| v.as_str()) {
                if artifact.get("type").and_then(|v| v.as_str()) != Some(artifact_type) {
                    return false;
                }
            }

            // Objective tag filter
            if let Some(objective_tag) = query_value.get("objectiveTag").and_then(|v| v.as_str()) {
                if let Some(tags) = artifact.get("objectiveTags").and_then(|v| v.as_array()) {
                    let has_tag = tags.iter().any(|t| t.as_str() == Some(objective_tag));
                    if !has_tag {
                        return false;
                    }
                } else {
                    return false;
                }
            }

            // Design pack ID filter
            if let Some(pack_id) = query_value.get("designPackId").and_then(|v| v.as_str()) {
                if artifact.get("designPackId").and_then(|v| v.as_str()) != Some(pack_id) {
                    return false;
                }
            }

            // Search text filter (title)
            if let Some(search_text) = query_value.get("searchText").and_then(|v| v.as_str()) {
                let search_lower = search_text.to_lowercase();
                if let Some(title) = artifact.get("title").and_then(|v| v.as_str()) {
                    if !title.to_lowercase().contains(&search_lower) {
                        return false;
                    }
                } else {
                    return false;
                }
            }

            true
        })
        .collect();

    serde_json::to_string(&filtered).map_err(|e| format!("Failed to serialize results: {}", e))
}

File: src/services/library-storage.ts
import { invoke } from "@tauri-apps/api/core";
import { isTauriContext } from "./tauri-bridge";
import type {
  LocalArtifact,
  ArtifactSearchQuery,
  LibraryIndex,
} from "@/types";

// ============================================
// Library Index Functions
// ============================================

/**
 * Get the library index (metadata for all artifacts)
 */
export async function getLibraryIndex(): Promise<LibraryIndex> {
  if (!isTauriContext()) {
    // Browser fallback - use localStorage
    const stored = localStorage.getItem("library-index");
    if (stored) return JSON.parse(stored);
    return {
      version: 1,
      lastUpdated: new Date().toISOString(),
      artifacts: [],
    };
  }

  const result = await invoke<string>("get_library_index");
  return JSON.parse(result);
}

/**
 * Save the library index
 */
export async function saveLibraryIndex(index: LibraryIndex): Promise<void> {
  if (!isTauriContext()) {
    // Browser fallback
    localStorage.setItem("library-index", JSON.stringify(index));
    return;
  }

  await invoke("save_library_index", {
    index: JSON.stringify(index),
  });
}

// ============================================
// Artifact Functions
// ============================================

/**
 * Get all artifacts (from index, without HTML content)
 */
export async function getArtifacts(): Promise<Omit<LocalArtifact, "htmlContent">[]> {
  const index = await getLibraryIndex();
  return index.artifacts;
}

/**
 * Get artifacts filtered by project ID
 */
export async function getArtifactsByProject(projectId: string): Promise<Omit<LocalArtifact, "htmlContent">[]> {
  const index = await getLibraryIndex();
  return index.artifacts.filter((a) => a.projectId === projectId);
}

/**
 * Get a specific artifact by ID (includes HTML content)
 */
export async function getArtifact(artifactId: string): Promise<LocalArtifact | null> {
  if (!isTauriContext()) {
    // Browser fallback - store full artifacts in localStorage
    const stored = localStorage.getItem(`artifact-${artifactId}`);
    if (stored) return JSON.parse(stored);
    return null;
  }

  try {
    const result = await invoke<string>("get_artifact", { artifactId });
    return JSON.parse(result);
  } catch {
    return null;
  }
}

/**
 * Save an artifact
 */
export async function saveArtifact(artifact: LocalArtifact): Promise<void> {
  if (!isTauriContext()) {
    // Browser fallback
    // Save full artifact separately
    localStorage.setItem(`artifact-${artifact.artifactId}`, JSON.stringify(artifact));

    // Update index (without HTML content)
    const index = await getLibraryIndex();
    const indexEntry: Omit<LocalArtifact, "htmlContent"> = {
      artifactId: artifact.artifactId,
      projectId: artifact.projectId,
      jobId: artifact.jobId,
      type: artifact.type,
      title: artifact.title,
      grade: artifact.grade,
      subject: artifact.subject,
      objectiveTags: artifact.objectiveTags,
      designPackId: artifact.designPackId,
      filePath: artifact.filePath,
      createdAt: artifact.createdAt,
    };

    // Remove existing entry with same ID
    index.artifacts = index.artifacts.filter((a) => a.artifactId !== artifact.artifactId);
    index.artifacts.push(indexEntry as LocalArtifact);
    index.lastUpdated = new Date().toISOString();

    await saveLibraryIndex(index);
    return;
  }

  await invoke("save_artifact", {
    artifact: JSON.stringify(artifact),
  });
}

/**
 * Delete an artifact
 */
export async function deleteArtifact(artifactId: string): Promise<void> {
  if (!isTauriContext()) {
    // Browser fallback
    localStorage.removeItem(`artifact-${artifactId}`);

    // Update index
    const index = await getLibraryIndex();
    index.artifacts = index.artifacts.filter((a) => a.artifactId !== artifactId);
    index.lastUpdated = new Date().toISOString();
    await saveLibraryIndex(index);
    return;
  }

  await invoke("delete_artifact", { artifactId });
}

/**
 * Search artifacts with filters
 */
export async function searchArtifacts(query: ArtifactSearchQuery): Promise<Omit<LocalArtifact, "htmlContent">[]> {
  if (!isTauriContext()) {
    // Browser fallback - filter locally
    const index = await getLibraryIndex();
    let results = index.artifacts;

    if (query.projectId) {
      results = results.filter((a) => a.projectId === query.projectId);
    }
    if (query.grade) {
      results = results.filter((a) => a.grade === query.grade);
    }
    if (query.subject) {
      results = results.filter((a) => a.subject === query.subject);
    }
    if (query.type) {
      results = results.filter((a) => a.type === query.type);
    }
    if (query.objectiveTag) {
      results = results.filter((a) => a.objectiveTags.includes(query.objectiveTag!));
    }
    if (query.designPackId) {
      results = results.filter((a) => a.designPackId === query.designPackId);
    }
    if (query.searchText) {
      const searchLower = query.searchText.toLowerCase();
      results = results.filter((a) => a.title.toLowerCase().includes(searchLower));
    }

    return results;
  }

  const result = await invoke<string>("search_artifacts", {
    query: JSON.stringify(query),
  });
  return JSON.parse(result);
}

/**
 * Update the objective tags on an artifact
 */
export async function updateArtifactTags(artifactId: string, tags: string[]): Promise<void> {
  // Update the full artifact
  const artifact = await getArtifact(artifactId);
  if (!artifact) throw new Error(`Artifact ${artifactId} not found`);

  artifact.objectiveTags = tags;
  await saveArtifact(artifact);
}

/**
 * Get all artifacts from the same generation job
 */
export async function getArtifactsByJob(jobId: string): Promise<LocalArtifact[]> {
  const index = await getLibraryIndex();
  const jobArtifacts = index.artifacts.filter((a) => a.jobId === jobId);

  // Load full content for each artifact
  const fullArtifacts: LocalArtifact[] = [];
  for (const meta of jobArtifacts) {
    const full = await getArtifact(meta.artifactId);
    if (full) fullArtifacts.push(full);
  }
  return fullArtifacts;
}

/**
 * Create multiple artifacts from a generation result
 */
export async function saveArtifactsFromGeneration(
  projectId: string,
  jobId: string,
  grade: string,
  subject: string,
  title: string,
  objectiveTags: string[],
  designPackId: string | undefined,
  contents: {
    studentPage?: string;
    teacherScript?: string;
    answerKey?: string;
    lessonPlan?: string;
  }
): Promise<LocalArtifact[]> {
  const now = new Date().toISOString();
  const artifacts: LocalArtifact[] = [];

  if (contents.studentPage) {
    const artifact: LocalArtifact = {
      artifactId: crypto.randomUUID(),
      projectId,
      jobId,
      type: "student_page",
      title: `${title} - Student Page`,
      htmlContent: contents.studentPage,
      grade: grade as LocalArtifact["grade"],
      subject,
      objectiveTags,
      designPackId,
      createdAt: now,
    };
    await saveArtifact(artifact);
    artifacts.push(artifact);
  }

  if (contents.teacherScript) {
    const artifact: LocalArtifact = {
      artifactId: crypto.randomUUID(),
      projectId,
      jobId,
      type: "teacher_script",
      title: `${title} - Teacher Script`,
      htmlContent: contents.teacherScript,
      grade: grade as LocalArtifact["grade"],
      subject,
      objectiveTags,
      designPackId,
      createdAt: now,
    };
    await saveArtifact(artifact);
    artifacts.push(artifact);
  }

  if (contents.answerKey) {
    const artifact: LocalArtifact = {
      artifactId: crypto.randomUUID(),
      projectId,
      jobId,
      type: "answer_key",
      title: `${title} - Answer Key`,
      htmlContent: contents.answerKey,
      grade: grade as LocalArtifact["grade"],
      subject,
      objectiveTags,
      designPackId,
      createdAt: now,
    };
    await saveArtifact(artifact);
    artifacts.push(artifact);
  }

  if (contents.lessonPlan) {
    const artifact: LocalArtifact = {
      artifactId: crypto.randomUUID(),
      projectId,
      jobId,
      type: "lesson_plan",
      title: `${title} - Lesson Plan`,
      htmlContent: contents.lessonPlan,
      grade: grade as LocalArtifact["grade"],
      subject,
      objectiveTags,
      designPackId,
      createdAt: now,
    };
    await saveArtifact(artifact);
    artifacts.push(artifact);
  }

  return artifacts;
}
